bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/BigtableOptionsFactory.java

```diff
@@ -167,50 +191,22 @@ public static BigtableOptions fromConfiguration(Configuration configuration) thr
     bigtableOptionsBuilder.setCluster(cluster);
     LOG.debug("Cluster %s", cluster);
 
-    // TODO(sduskis): InetAddress.getByName is expensive.  This can be parallelized to speed up
-    // startup.
-
     String overrideIp = configuration.get(IP_OVERRIDE_KEY);
-    InetAddress overrideIpAddress = null;
-    if (!Strings.isNullOrEmpty(overrideIp)) {
+    final InetAddress overrideIpAddress =
+        !Strings.isNullOrEmpty(overrideIp) ? InetAddress.getByName(overrideIp) : null;
+    if (overrideIp != null) {
       LOG.debug("Using override IP address %s", overrideIp);
-      overrideIpAddress = InetAddress.getByName(overrideIp);
     }
 
-    String dataHost = configuration.get(BIGTABLE_HOST_KEY, BIGTABLE_HOST_DEFAULT);
-    Preconditions.checkArgument(
-        !Strings.isNullOrEmpty(dataHost),
-        String.format("API Data endpoint host must be supplied via %s", BIGTABLE_HOST_KEY));
-    if (overrideIpAddress == null) {
-      LOG.debug("Data endpoint host %s", dataHost);
-      bigtableOptionsBuilder.setDataHost(InetAddress.getByName(dataHost));
-    } else {
-      LOG.debug("Data endpoint host %s. Using override IP address.", dataHost);
-      bigtableOptionsBuilder.setDataHost(
-          InetAddress.getByAddress(dataHost, overrideIpAddress.getAddress()));
-    }
+    // It could take a while for the InetAddresses to resolve.  Resolve them asynchronously.
+    bigtableOptionsBuilder.setDataHost(getHost(configuration, overrideIpAddress,
+      BIGTABLE_HOST_KEY, BIGTABLE_HOST_DEFAULT, "API Data"));
 
-    String tableAdminHost = configuration.get(
-        BIGTABLE_TABLE_ADMIN_HOST_KEY, BIGTABLE_TABLE_ADMIN_HOST_DEFAULT);
-    if (overrideIpAddress == null) {
-      LOG.debug("Table admin endpoint host %s", tableAdminHost);
-      bigtableOptionsBuilder.setTableAdminHost(InetAddress.getByName(tableAdminHost));
-    } else {
-      LOG.debug("Table admin endpoint host %s. Using override IP address.", tableAdminHost);
-      bigtableOptionsBuilder.setTableAdminHost(
-          InetAddress.getByAddress(tableAdminHost, overrideIpAddress.getAddress()));
-    }
+    bigtableOptionsBuilder.setTableAdminHost(getHost(configuration, overrideIpAddress,
+      BIGTABLE_TABLE_ADMIN_HOST_KEY, BIGTABLE_TABLE_ADMIN_HOST_DEFAULT, "Table admin"));
```

Comment:
"API Data" and "Table Admin" doesn't seem consistent. "(Data|Table admin|...) endpoint"?